tests/support/jupyter.py

```diff
@@ -0,0 +1,432 @@
+"""Python code to get the node on which the lab is running."""
+
+
+class JupyterAction(Enum):
+    LOGIN = "login"
+    HOME = "home"
+    HUB = "hub"
+    USER = "user"
+    PROGRESS = "progress"
+    SPAWN = "spawn"
+    SPAWN_PENDING = "spawn_pending"
+    LAB = "lab"
+    DELETE_LAB = "delete_lab"
+    CREATE_SESSION = "create_session"
+    DELETE_SESSION = "delete_session"
+
+
+class JupyterState(Enum):
+    LOGGED_OUT = "logged out"
+    LOGGED_IN = "logged in"
+    SPAWN_PENDING = "spawn pending"
+    LAB_RUNNING = "lab running"
+
+
+def url_for_path(route: str) -> str:
+    """Construct a URL for JupyterHub/Proxy."""
+    return f"{config.environment_url}/nb/{route}"
+
+
+def url_for_pattern(route: str) -> str:
+    """Construct a URL for JupyterHub/Proxy."""
+    prefix = re.escape(f"{config.environment_url}/nb/")
+    return prefix + route
+
+
+class MockJupyter:
+    """A mock Jupyter state machine.
+
+    This should be invoked via mocked HTTP calls so that tests can simulate
+    making REST calls to the real JupyterHub/Lab.
+    """
+
+    def __init__(self) -> None:
+        self.sessions: Dict[str, JupyterLabSession] = {}
+        self.state: Dict[str, JupyterState] = {}
+        self.delete_immediate = True
+        self.spawn_timeout = False
+        self._delete_at: Dict[str, Optional[datetime.datetime]] = {}
+        self._fail: Dict[str, Dict[JupyterAction, bool]] = {}
+
+    def fail(self, user: str, action: JupyterAction) -> None:
+        """Configure the given action to fail for the given user."""
+        if user not in self._fail:
+            self._fail[user] = {}
+        self._fail[user][action] = True
+
+    def login(self, request: httpx.Request) -> httpx.Response:
+        """Mock the JupyterHub log in response."""
+        user = self._get_user(request.headers["Authorization"])
+        if JupyterAction.LOGIN in self._fail.get(user, {}):
+            return httpx.Response(500, request=request)
+        state = self.state.get(user, JupyterState.LOGGED_OUT)
+        if state == JupyterState.LOGGED_OUT:
+            self.state[user] = JupyterState.LOGGED_IN
+        return httpx.Response(200, request=request)
+
+    def user(self, request: httpx.Request) -> httpx.Response:
+        user = self._get_user(request.headers["Authorization"])
+        if JupyterAction.USER in self._fail.get(user, {}):
+            return httpx.Response(500, request=request)
+        assert str(request.url).endswith(f"/hub/api/users/{user}")
+        state = self.state.get(user, JupyterState.LOGGED_OUT)
+        if state == JupyterState.SPAWN_PENDING:
+            server = {"name": "", "pending": "spawn", "ready": False}
+            body = {"name": user, "servers": {"": server}}
+        elif state == JupyterState.LAB_RUNNING:
+            delete_at = self._delete_at.get(user)
+            if (
+                delete_at
+                and (datetime.datetime.now(tz=datetime.timezone.utc))
+                > delete_at
+            ):
```

Comment:
Nit: I think `datetime.datetime` is pretty ugly and verbose, so I always import `datetime`, `timedelta`, and `timezone` from `datetime` instead.

Nit: Pulling `now = datetime.now(tz=timezone.utc)` out of the conditional will avoid the ugly and hard-to-read wrapping. It's slightly slower but it's the test suite so there's no reason to care.